src/pks/flow/preferential_pk.cc

```diff
@@ -0,0 +1,597 @@
+/* -*-  mode: c++; indent-tabs-mode: nil -*- */
+
+/* -------------------------------------------------------------------------
+This is the flow component of the Amanzi code.
+License: BSD
+Authors: Daniil Svyatsky(dasvyat@gmail.com)
+------------------------------------------------------------------------- */
+#include "boost/math/special_functions/fpclassify.hpp"
+
+#include "boost/algorithm/string/predicate.hpp"
+
+#include "Epetra_Import.h"
+
+#include "flow_bc_factory.hh"
+
+#include "Point.hh"
+
+#include "upwind_cell_centered.hh"
+#include "upwind_arithmetic_mean.hh"
+#include "upwind_total_flux.hh"
+#include "upwind_gravity_flux.hh"
+
+#include "CompositeVectorFunction.hh"
+#include "CompositeVectorFunctionFactory.hh"
+
+#include "predictor_delegate_bc_flux.hh"
+#include "wrm_evaluator.hh"
+#include "rel_perm_evaluator.hh"
+#include "richards_water_content_evaluator.hh"
+#include "OperatorDefs.hh"
+#include "BoundaryFlux.hh"
+#include "pk_helpers.hh"
+
+#include "preferential.hh"
+
+#define DEBUG_RES_FLAG 0
+
+
+namespace Amanzi {
+namespace Flow {
+
+// -------------------------------------------------------------
+// Constructor
+// -------------------------------------------------------------
+
+Preferential::Preferential(Teuchos::ParameterList& pk_tree,
+                   const Teuchos::RCP<Teuchos::ParameterList>& glist,
+                   const Teuchos::RCP<State>& S,
+                   const Teuchos::RCP<TreeVector>& solution) :
+    PK(pk_tree, glist,  S, solution),
+    Richards(pk_tree, glist,  S, solution)
+{
+    coef_grav_key_ = Keys::readKey(*plist_, domain_, "gravity conductivity", "gravity_relative_permeability");
+    // set up an additional primary variable evaluator for flux
+    Teuchos::ParameterList& pv_sublist = S->GetEvaluatorList(flux_key_);
+    pv_sublist.set("field evaluator type", "primary variable");
+}
+
+// -------------------------------------------------------------
+// Setup data
+// -------------------------------------------------------------
+void Preferential::Setup()
+{
+  PK_PhysicalBDF_Default::Setup();
+  SetupPreferentialFlow_();
+  SetupDiscretization_();
+  SetupPhysicalEvaluators_();
```

Comment:
I'd like to take a look at what is where here.   It looks like a lot of the code in SetupPreferentialFlow is already in Richards.  Surely we can avoid duplicating that?

Reply:
resolved on ecoon branch